backend/functions/src/v2/course/survey/functions.ts

```diff
@@ -0,0 +1,89 @@
+import { Request, Response } from "express";
+import { SurveyQuestion, IIndex } from "../../../common/types";
+import { getCourseRefByDocId } from "../../../common/utils";
+
+export const getSurvey = async (req: Request, res: Response) => {
+	const { courseId, studentId } = req.params;
+
+	const courseDocRef = (await getCourseRefByDocId(courseId).catch((err) => {
+		return res.status(404).send(err.toString());
+	})) as FirebaseFirestore.DocumentReference;
+	const surveyColRef = courseDocRef.collection("survey");
+
+	const surveyQuery = await surveyColRef
+		.where("studentId", "==", studentId)
+		.limit(1)
+		.get();
+
+	if (surveyQuery.empty) {
+		return res
+			.status(404)
+			.send(`Survey for ${studentId} doesn't exist under course ${courseId}`);
+	}
+	return res.status(200).send(surveyQuery.docs[0].data());
+};
+
+export const postSurvey = async (req: Request, res: Response) => {
+	const courseId = req.params.courseId;
+	const { studentId, fullName, surveyResponse } = req.body;
+	// Check essential fields in request body
+	const isInvalidBody = [courseId, studentId, fullName, surveyResponse].some(
+		(v) => v === undefined
+	);
+	if (isInvalidBody) {
+		return res.status(400).send(
+			// eslint-disable-next-line max-len
+			"Invalid request: the required fields are: studentId, fullname, surveyResponse."
+		);
+	}
+	// Get question "schema" from course doc
+	const courseDocRef = (await getCourseRefByDocId(courseId).catch((err) => {
+		return res.status(404).send(err.toString());
+	})) as FirebaseFirestore.DocumentReference;
+	const courseDocSnapshot = await courseDocRef.get();
+	const question = courseDocSnapshot.get("question") as SurveyQuestion[];
+	// Validate survey response against schema
+	if (isSurveyResponseInvalid(surveyResponse, question)) {
+		return res.status(400).send("Invalid survey response(s)");
+	}
+	// Check for existing survey response under studentId
+	const surveyColRef = courseDocRef.collection("survey");
+	const surveyQuery = await surveyColRef
+		.where("studentId", "==", studentId)
+		.limit(1)
+		.get();
+	if (!surveyQuery.empty) {
+		return res.status(409).send(
+			// eslint-disable-next-line max-len
+			`Survey response for ${studentId} already exists! Use the PATCH endpoint instead.`
+		);
+	}
+	// Write valid surveyResponse to survey subcollection
+	const newSurveyDocRef = await surveyColRef.add({
+		studentId,
```

Comment:
here too. but like i said ill leave this discussion to you and @XiaochenSeanYu. just tell me what the two of you decide on.

Reply:
@shichong-97  Ohh so as per my PR message above, this is not the finalized version. I wanted to break the next two parts into reasonably sized diffs, but apologies for the confusion anyway!

Reply:
Small PRs are nice